esp-hal/src/lib.rs

```diff
@@ -494,68 +494,92 @@ pub struct Config {
-        WatchdogStatus::Enabled(duration) => {
-            let mut timg1_wd = crate::timer::timg::Wdt::<self::peripherals::TIMG1>::new();
-            timg1_wd.enable();
-            timg1_wd.set_timeout(crate::timer::timg::MwdtStage::Stage0, duration);
+        match config.watchdog.timg0 {
+            WatchdogStatus::Enabled(duration) => {
+                let mut timg0_wd = crate::timer::timg::Wdt::<self::peripherals::TIMG0>::new();
+                timg0_wd.enable();
+                timg0_wd.set_timeout(crate::timer::timg::MwdtStage::Stage0, duration);
+            }
+            WatchdogStatus::Disabled => {
+                crate::timer::timg::Wdt::<self::peripherals::TIMG0>::new().disable();
+            }
         }
-        WatchdogStatus::Disabled => {
-            crate::timer::timg::Wdt::<self::peripherals::TIMG1>::new().disable();
+
+        #[cfg(timg1)]
+        match config.watchdog.timg1 {
+            WatchdogStatus::Enabled(duration) => {
+                let mut timg1_wd = crate::timer::timg::Wdt::<self::peripherals::TIMG1>::new();
+                timg1_wd.enable();
+                timg1_wd.set_timeout(crate::timer::timg::MwdtStage::Stage0, duration);
+            }
+            WatchdogStatus::Disabled => {
+                crate::timer::timg::Wdt::<self::peripherals::TIMG1>::new().disable();
+            }
         }
-    }
 
-    Clocks::init(config.cpu_clock);
+        Clocks::init(config.cpu_clock);
 
-    #[cfg(esp32)]
-    crate::time::time_init();
+        #[cfg(esp32)]
+        crate::time::time_init();
 
-    crate::gpio::bind_default_interrupt_handler();
+        crate::gpio::bind_default_interrupt_handler();
 
-    #[cfg(any(feature = "quad-psram", feature = "octal-psram"))]
-    crate::psram::init_psram(config.psram);
+        #[cfg(any(feature = "quad-psram", feature = "octal-psram"))]
+        crate::psram::init_psram(config.psram);
+    }
+}
+
+/// Initialize the system.
+///
+/// This function sets up the CPU clock and watchdog, then, returns the
+/// peripherals and clocks.
+pub fn init(config: Config) -> Peripherals {
+    system::disable_peripherals();
+
+    let mut peripherals = if let Some(p) = Peripherals::take() {
+        p
+    } else {
+        panic!("init called more than once!");
+    };
+
+    init_internal(Some(&mut peripherals), config);
 
     peripherals
 }
+
+/// Initializes the system only if necessary, then returns the peripherals.
+///
+/// If [`init`] has already been executed, simply steals and returns the
+/// peripherals. (This check uses [`AtomicBool`](core::sync::atomic::AtomicBool)
+/// to be as thread-safe as possible)
+///
+/// Otherwise, the initialization process is executed (and then the peripherals
+/// are stolen and handed off to you).
+pub unsafe fn try_init(config: Config) -> Peripherals {
+    let mut peripherals = Peripherals::take();
+    init_internal(peripherals.as_mut(), config);
+    peripherals.unwrap_or_else(|| Peripherals::steal())
```

Comment:
```suggestion
    Peripherals::steal()
```

Reply:
I'm a bit confused by this suggestion. Why create a second instance of `Peripherals` here when it's not necessary (or will the compiler abstract that away regardless, or some other thing I'm not thinking of)?

Reply:
It does the same thing as the original line, but with less code. If you're already stealing as a fallback, might as well just steal anyway and you'll avoid code that reads `a.unwrap_or(a)`.